File: spinnerDialog/src/main/java/in/galaxyofandroid/spinnerdialog/SpinnerDialog.java
package in.galaxyofandroid.spinnerdialog;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Application;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.*;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.Loader;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by Md Farhan Raja on 2/23/2017
 */

public final class SpinnerDialog<E extends Parcelable> extends DialogFragment implements LoaderManager.LoaderCallbacks<List<E>> {

    public static <E extends Parcelable> SpinnerDialog<E> create(@StringRes int titleRes, ItemManager<E> itemManager,
                                                                 @StringRes int emptyTextRes,
                                                                 @DrawableRes int emptyIconRes,
                                                                 @StringRes int cancelRes) {
        if (titleRes <= 0) throw new IllegalArgumentException("invalid title resource");
        if (emptyTextRes <= 0) throw new IllegalArgumentException("invalid 'empty text' resource");

        SpinnerDialog<E> dialog = new SpinnerDialog<>();

        Bundle args = new Bundle(2);
        args.putInt("titleRes", titleRes);
        args.putParcelable("item manager", required(itemManager, "item manager"));
        args.putInt("emptyTextRes", emptyTextRes);
        args.putInt("emptyIconRes", emptyIconRes);
        args.putInt("cancelRes", cancelRes);
        dialog.setArguments(args);

        return dialog;
    }

    private String filter;
    private SpinnerDialogAdapter<E> adapter;
    private View progress;
    private TextView empty;

    public SpinnerDialog withWindowAnimations(@StyleRes int windowAnimations) {
        getArguments().putInt("animations", windowAnimations);
        return this;
    }

    public void show(FragmentManager manager, Fragment caller, int requestCode) {
        setTargetFragment(required(caller, "caller fragment"), requestCode);
        show(required(manager, "fragment manager"), null);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getLoaderManager().initLoader(0, null, this);
    }

    @NonNull @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final String title;
        final ItemManager<E> itemManager;
        final String emptyText;
        final Drawable emptyIcon;
        final int windowAnimations;
        final int cancelRes;
        {
            Bundle args = getArguments();
            title = getString(args.getInt("titleRes"));
            itemManager = required(args.<ItemManager<E>>getParcelable("item manager"), "item manager");
            emptyText = getString(args.getInt("emptyTextRes"));
            int emptyIconRes = args.getInt("emptyIconRes");
            emptyIcon = emptyIconRes > 0 ? ContextCompat.getDrawable(getActivity(), emptyIconRes) : null;
            windowAnimations = args.getInt("animations", -1);
            int _cancelRes = args.getInt("cancelRes", -1);
            if (_cancelRes > 0) cancelRes = _cancelRes;
            else cancelRes = android.R.string.cancel;
        }

        Activity activity = getActivity();
        View v = activity.getLayoutInflater().inflate(R.layout.dialog_layout, null, false);

        final RecyclerView recyclerView = (RecyclerView) v.findViewById(R.id.list);
        final EditText searchBox = (EditText) v.findViewById(R.id.searchBox);
        adapter = new SpinnerDialogAdapter<>(getActivity(), itemManager);
        recyclerView.setLayoutManager(new LinearLayoutManager(activity));
        recyclerView.setAdapter(adapter);
        recyclerView.addItemDecoration(new DividerItemDecoration(activity, DividerItemDecoration.VERTICAL));
        progress = v.findViewById(R.id.progress);
        empty = (TextView) v.findViewById(R.id.empty);
        empty.setText(emptyText);
        empty.setCompoundDrawablesWithIntrinsicBounds(emptyIcon, null, null, null);
        final AlertDialog alertDialog =
                new AlertDialog.Builder(activity)
                        .setTitle(title)
                        .setView(v)
                        .setNegativeButton(cancelRes, null)
                        .create();

        if (windowAnimations > 0) {
            alertDialog.getWindow().getAttributes().windowAnimations = windowAnimations;
        }

        adapter.setListener(new SpinnerDialogAdapter.OnItemClickListener<E>() {
            @Override public void onItemClick(E e) {
                Intent data = new Intent();
                data.putExtra("item", e);
                getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, data);
                alertDialog.dismiss();
            }
            @Override public void onItemBound(int position) {
                if (((PagedLoader) getLoaderManager().getLoader(0)).onItemBound(position)) {
                    progress.setVisibility(View.VISIBLE);
                }
            }
        });

        searchBox.addTextChangedListener(new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
            @Override public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
            @Override public void afterTextChanged(Editable editable) {
                progress.setVisibility(View.VISIBLE);
                filter = searchBox.getText().toString();
                getLoaderManager().restartLoader(0, null, SpinnerDialog.this);
            }
        });
        return alertDialog;
    }

    private static <T> T required(T t, String tag) {
        if (t == null) throw new NullPointerException(tag + " is required");
        return t;
    }

    @Override
    public Loader<List<E>> onCreateLoader(int id, Bundle args) {
        return createLoader(getActivity(), filter, getArguments().<ItemManager<E>>getParcelable("item manager"));
    }

    private static <E extends Parcelable> Loader<List<E>> createLoader(
            Context context, final @Nullable String filter, final ItemManager<E> manager) {
        return new PagedLoader<>(context, filter, manager);
    }

    @Override
    public void onLoadFinished(Loader<List<E>> loader, List<E> data) {
        adapter.setData(data);
        progress.setVisibility(View.GONE);
        empty.setVisibility(data.isEmpty() ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onLoaderReset(Loader<List<E>> loader) {
    }

    private static final class PagedLoader<E extends Parcelable> extends AsyncTaskLoader<List<E>> {

        private final String filter;
        private final ItemManager<E> itemManager;

        PagedLoader(Context context, String filter, ItemManager<E> itemManager) {
            super(context);
            this.filter = filter;
            this.itemManager = itemManager;
        }

        private List<E> list = Collections.emptyList();
        private volatile boolean loading = false;

        @Override protected void onStartLoading() {
            int total = itemManager.getTotal((Application) getContext(), filter);
            if (list.isEmpty() && total != 0 || total > list.size() && takeContentChanged())
                forceLoad(); // total size is unknown or greater than available size
            else
                deliverResult(list); // we've got it all
        }

        @Override public List<E> loadInBackground() {
            List<E> loaded = itemManager.load((Application) getContext(), filter, list.size());
            List<E> aList = new ArrayList<>(list.size() + loaded.size());
            aList.addAll(list);
            aList.addAll(loaded);
            loading = false;
            return list = aList; // we can't just add, Loader won't deliver the same object.
        }

        boolean onItemBound(int position) {
            int size = list.size();
            if (position > size - 3 && !loading) {
                int total = itemManager.getTotal((Application) getContext(), filter);
                if (total < 0 || total > size) {
                    loading = true;
                    onContentChanged();
                    return true;
                }
            }
            return false;
        }
    }
}

File: spinnerDialog/src/main/java/in/galaxyofandroid/spinnerdialog/SpinnerDialog.kt
package `in`.galaxyofandroid.spinnerdialog

import android.app.Activity
import android.app.AlertDialog
import android.app.Application
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.support.annotation.DrawableRes
import android.support.annotation.StringRes
import android.support.annotation.StyleRes
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.LoaderManager
import android.support.v4.content.AsyncTaskLoader
import android.support.v4.content.ContextCompat
import android.support.v4.content.Loader
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import java.util.*

/**
 * Created by Md Farhan Raja on 2/23/2017
 */

class SpinnerDialog<E : Parcelable> : DialogFragment(), LoaderManager.LoaderCallbacks<Either<Throwable, List<E>>> {

    private lateinit var adapter: SpinnerDialogAdapter<E>
    private lateinit var progress: View
    private lateinit var empty: TextView
    private lateinit var searchBox: TextView

    fun withWindowAnimations(@StyleRes windowAnimations: Int): SpinnerDialog<*> {
        arguments.putInt("animations", windowAnimations)
        return this
    }

    fun show(manager: FragmentManager, caller: Fragment, requestCode: Int) {
        setTargetFragment(caller, requestCode)
        show(manager, null)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        loaderManager.initLoader(0, null, this)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = arguments
        val title = getString(args.getInt("titleRes"))
        val itemManager = args.getParcelable<ItemManager<E>>("item manager")
        val emptyText = getString(args.getInt("emptyTextRes"))
        val emptyIcon = args.getInt("emptyIconRes").let { if (it > 0) ContextCompat.getDrawable(activity, it) else null }
        val windowAnimations = args.getInt("animations", -1)
        val cancelRes = args.getInt("cancelRes", -1).let { if (it > 0) it else android.R.string.cancel }

        val activity = activity
        val v = activity.layoutInflater.inflate(R.layout.dialog_layout, null, false)

        val recyclerView = v.findViewById(R.id.list) as RecyclerView
        searchBox = v.findViewById(R.id.searchBox) as EditText
        adapter = SpinnerDialogAdapter(getActivity(), itemManager)
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.adapter = adapter
        recyclerView.addItemDecoration(DividerItemDecoration(activity, DividerItemDecoration.VERTICAL))
        progress = v.findViewById(R.id.progress)
        empty = v.findViewById(R.id.empty) as TextView
        empty.text = emptyText
        empty.setCompoundDrawablesWithIntrinsicBounds(emptyIcon, null, null, null)

        val alertDialog = AlertDialog.Builder(activity)
                .setTitle(title)
                .setView(v)
                .setNegativeButton(cancelRes, null)
                .create()

        if (windowAnimations > 0) {
            alertDialog.window.attributes.windowAnimations = windowAnimations
        }

        adapter.listener = object : SpinnerDialogAdapter.OnItemClickListener<E> {
            override fun onItemClick(e: E) {
                targetFragment.onActivityResult(
                        targetRequestCode, Activity.RESULT_OK, Intent().also { it.putExtra("item", e) })
                alertDialog.dismiss()
            }

            override fun onItemBound(position: Int) {
                if ((loaderManager.getLoader<Any>(0) as PagedLoader<*>).onItemBound(position)) {
                    progress.visibility = View.VISIBLE
                }
            }
        }
        return alertDialog
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Either<Throwable, List<E>>> {
        searchBox.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                progress.visibility = View.VISIBLE
                val newFilter = searchBox.text.toString().let { if (it.isBlank()) null else it }
                val filter = (loaderManager.getLoader<Nothing>(0) as PagedLoader<*>).filter
                if (filter != newFilter) {
                    loaderManager.restartLoader(0, null, this@SpinnerDialog)
                }
            }
        })

        return PagedLoader(activity,
                searchBox.text.let { if (it.isBlank()) null else it.toString() },
                arguments.getParcelable("item manager"))
    }

    override fun onLoadFinished(loader: Loader<Either<Throwable, List<E>>>?, data: Either<Throwable, List<E>>) =
            data.let({
                Toast.makeText(activity, it.message ?: it.toString(), Toast.LENGTH_LONG).show() // todo: proper error handling
            }, {
                adapter.setData(it)
                progress.visibility = View.GONE
                empty.visibility = if (it.isEmpty()) View.VISIBLE else View.GONE
            })

    override fun onLoaderReset(loader: Loader<Either<Throwable, List<E>>>?) {}

    private class PagedLoader<E : Parcelable>
    internal constructor(
            context: Context,
            internal val filter: String?,
            private val itemManager: ItemManager<E>
    ) : AsyncTaskLoader<Either<Throwable, List<E>>>(context) {

        private var list: List<E>? = null
        @Volatile private var loading = false
        private var currentPage = 0
        private var lastPage = Int.MAX_VALUE

        override fun onStartLoading() {
            val list = list
            if (!loading && (list == null || currentPage < lastPage))
                forceLoad() // if not loading yet, and if there's no or insufficient data
            else if (list != null)
                deliverResult(Right(list)) // we have some data
        }

        override fun loadInBackground(): Either<Throwable, List<E>> =
                itemManager.load(context as Application, filter, currentPage + 1).map({ (loadedList, lastPage) ->
                    val aList = ArrayList<E>(list?.size ?: 0 + loadedList.size)
                    list?.let { aList.addAll(it) }
                    aList.addAll(loadedList)
                    loading = false
                    list = aList
                    currentPage++
                    this.lastPage = lastPage
                    aList // we can't just add, Loader won't deliver the same object.
                })

        internal fun onItemBound(position: Int): Boolean {
            val size = list!!.size
            if (position > size - 3 && !loading && currentPage < lastPage) {
                loading = true
                forceLoad()
                return true
            }
            return false
        }
    }

    companion object {

        fun <E : Parcelable> create(@StringRes titleRes: Int, itemManager: ItemManager<E>,
                                    @StringRes emptyTextRes: Int,
                                    @DrawableRes emptyIconRes: Int,
                                    @StringRes cancelRes: Int): SpinnerDialog<E> {
            if (titleRes <= 0) throw IllegalArgumentException("invalid title resource")
            if (emptyTextRes <= 0) throw IllegalArgumentException("invalid 'empty text' resource")

            val dialog = SpinnerDialog<E>()

            val args = Bundle(5)
            args.putInt("titleRes", titleRes)
            args.putParcelable("item manager", itemManager)
            args.putInt("emptyTextRes", emptyTextRes)
            args.putInt("emptyIconRes", emptyIconRes)
            args.putInt("cancelRes", cancelRes)
            dialog.arguments = args

            return dialog
        }
    }
}
